python/pyarrow/tests/test_cuda.py

```diff
@@ -0,0 +1,551 @@
+
+    for (position, nbytes) in [
+        (size//2, size//2+1)
+    ]:
+        with pytest.raises(ValueError) as e_info:
+            put(position=position, nbytes=nbytes)
+            print('dbuf.size={}, buf.size={}, position={}, nbytes={}'
+                  .format(dbuf.size, buf.size, position, nbytes))
+        assert str(e_info.value).startswith(
+            'requested more to copy than available in device buffer')
+
+
+def test_BufferWriter():
+
+    def allocate(size):
+        cbuf = global_context.allocate(size)
+        writer = cuda.BufferWriter(cbuf)
+        return cbuf, writer
+
+    def test_writes(total_bytes, chunksize, buffer_size=0):
+        cbuf, writer = allocate(total_size)
+        arr, buf = make_random_buffer(size=total_size, target='host')
+
+        if buffer_size > 0:
+            writer.set_buffer_size(buffer_size)
+
+        position = writer.tell()
+        assert position == 0
+        writer.write(buf.slice(length=chunksize))
+        assert writer.tell() == chunksize
+        writer.seek(0)
+        position = writer.tell()
+        assert position == 0
+
+        while position < total_bytes:
+            bytes_to_write = min(chunksize, total_bytes - position)
+            writer.write(buf.slice(offset=position, length=bytes_to_write))
+            position += bytes_to_write
+
+        writer.flush()
+        assert cbuf.size == total_bytes
+        buf2 = cbuf.copy_to_host()
+        assert buf2.size == total_bytes
+        arr2 = np.frombuffer(buf2, dtype=np.uint8)
+        assert arr2.size == total_bytes
+        assert (arr2 == arr).all()
+
+    total_size, chunk_size = 1 << 16, 1000
+    test_writes(total_size, chunk_size)
+    test_writes(total_size, chunk_size, total_size // (1 << 4))
+
+    cbuf, writer = allocate(100)
+    writer.write(np.arange(100, dtype=np.uint8))
+    writer.writeat(50, np.arange(25, dtype=np.uint8))
+    writer.write(np.arange(25, dtype=np.uint8))
+    writer.flush()
+
+    arr = np.frombuffer(cbuf.copy_to_host(), np.uint8)
+    assert (arr[:50] == np.arange(50, dtype=np.uint8)).all()
+    assert (arr[50:75] == np.arange(25, dtype=np.uint8)).all()
+    assert (arr[75:] == np.arange(25, dtype=np.uint8)).all()
+
+
+def test_BufferWriter_edge_cases():
+    # edge cases:
+    size = 1000
+    cbuf = global_context.allocate(size)
+    writer = cuda.BufferWriter(cbuf)
+    arr, buf = make_random_buffer(size=size, target='host')
+
+    assert writer.buffer_size == 0
+    writer.set_buffer_size(100)
+    assert writer.buffer_size == 100
+
+    writer.write(buf.slice(length=0))
+    assert writer.tell() == 0
+
+    writer.write(buf.slice(length=10))
+    writer.set_buffer_size(200)
+    assert writer.buffer_size == 200
+    assert writer.num_bytes_buffered == 0
+
+    writer.write(buf.slice(offset=10, length=300))
+    assert writer.num_bytes_buffered == 0
+
+    writer.write(buf.slice(offset=310, length=200))
+    assert writer.num_bytes_buffered == 0
```

Comment:
This test comes from EdgeCases in cuda-test.cc. Added reference.
My interpretation of `num_bytes_buffered == 0` is that all preceding writes fulfill the buffer so that the `write` always triggers flush. Non-zero would indicate that flush is not triggered.